File: 백준/Gold/1715. 카드 정렬하기/카드 정렬하기.cc
#include <bits/stdc++.h>
using namespace std;

int n;
int result = 0;
priority_queue <int> heapq;

int main(void) {
	cin >> n;

	// 힙에 초기 카드 묶음을 모두 삽입
	for (int i = 0; i < n; i++) {
		int x;
		cin >> x;
		heapq.push(-x);           // 음수 형태로 넣어야 min-heap이 됨
	}

	// 원소가 1개 남을 때까지
	while (heapq.size() != 1) {
		int a = -heapq.top();
		heapq.pop();              // c++의 pop()은 리턴값이 없음!!

		int b = -heapq.top();
		heapq.pop();

		result += a + b;
		heapq.push(-(a + b));    // 음수 형태로 넣어야됨!!
	}
	cout << result << endl;;
}

File: 백준/Gold/1715. 카드 정렬하기/카드 정렬하기.py
import heapq
import sys
input = sys.stdin.readline

n = int(input())
arr = [int(input()) for _ in range(n)]

heapq.heapify(arr)   # 리스트릴 힙으로 변환(-> pop 시 min값 출력)

ans = 0

# 힙에 데이터 2개 이상 있어야 돌릴 수 있다.
while len(arr)-1 != 0:
    a = heapq.heappop(arr)   # min값 두 개를 추출
    b = heapq.heappop(arr)
    
    ans += a+b               # min값 두 개 더해주고
    heapq.heappush(arr, a+b) # 힙에 더한값 새롭게 넣는다 (= 기존 값과 비교가 필요하므로)
    
print(ans)